### PR TITLE
Fast log2 ceiling

### PR DESCRIPTION
**Pull request by HaoranYi**
#### Summary of Changes

Fixes #

### CODE REVIEW

File: bucket_map/src/index_entry.rs [hunk 1]
```diff
@@ -58,7 +58,8 @@ impl IndexEntry {
     }
 
     pub fn data_bucket_from_num_slots(num_slots: Slot) -> u64 {
-        (num_slots as f64).log2().ceil() as u64 // use int log here?
+        // Compute the ceiling of log2 for integer
+        64 - ((num_slots - 1).leading_zeros()) as u64
```
Comment (brooksprumo) [inline bucket_map/src/index_entry.rs:62]: Are there any concerns if `num_slots` is 0? Is wrapping correct? (I see there isn't a test case if `num_slots` is 0.)

If wrapping is desired, `num_slots` can be a [`Wrapping<Slot>`](https://doc.rust-lang.org/std/num/struct.Wrapping.html) to indicate desired behavior. If `num_slots` cannot be zero, then [`NonZeroU64`](https://doc.rust-lang.org/std/num/struct.NonZeroU64.html) can be used.

**Review by brooksprumo — COMMENTED** [review]
I like this. Intuitively I imagine this is a performance improvement; do you happen to have any measured numbers too?

**Comment by HaoranYi** [comment]
```
 pub fn data_bucket_from_num_slots_float(num_slots: u64) -> u64 {
        (num_slots as f64).log2().ceil() as u64 
 }

 pub fn data_bucket_from_num_slots_int(num_slots: u64) -> u64 {
        64 - ((num_slots - 1).leading_zeros()) as u64
 }
 ```
 
 generated code
 ```
 example::data_bucket_from_num_slots_float:
        push    rax
        movq    xmm1, rdi
        punpckldq       xmm1, xmmword ptr [rip + .LCPI0_0]
        subpd   xmm1, xmmword ptr [rip + .LCPI0_1]
        movapd  xmm0, xmm1
        unpckhpd        xmm0, xmm1
        addsd   xmm0, xmm1
        call    qword ptr [rip + log2@GOTPCREL]
        call    qword ptr [rip + ceil@GOTPCREL]
        cvttsd2si       rax, xmm0
        mov     rcx, rax
        sar     rcx, 63
        movapd  xmm1, xmm0
        subsd   xmm1, qword ptr [rip + .LCPI0_2]
        cvttsd2si       rdx, xmm1
        and     rdx, rcx
        or      rdx, rax
        xor     ecx, ecx
        xorpd   xmm1, xmm1
        ucomisd xmm0, xmm1
        cmovae  rcx, rdx
        ucomisd xmm0, qword ptr [rip + .LCPI0_3]
        mov     rax, -1
        cmovbe  rax, rcx
        pop     rcx
        ret

example::data_bucket_from_num_slots_int:
        mov     eax, 64
        mov     ecx, 64
        add     rdi, -1
        je      .LBB1_2
        bsr     rcx, rdi
        xor     rcx, 63
.LBB1_2:
        sub     rax, rcx
        ret
  ```

File: bucket_map/src/index_entry.rs (hunk 1 above)
Comment (HaoranYi) [reply bucket_map/src/index_entry.rs:62]: Good question. I think *num_slots* is slice size, should always be above 0? @jeffwashington 
If num_slots is zero, `(num_slots as f64).log2().ceil() as u64` will return 0. The new implementation will panic. 

If num_slots must be above zero, we could assert here
https://github.com/solana-labs/solana/blob/d8c45a69c3d0427898883ae03f320516d886f008/bucket_map/src/bucket.rs#L245

File: bucket_map/src/index_entry.rs [hunk 2]
```diff
@@ -153,4 +154,23 @@ mod tests {
         let mut index = IndexEntry::new(Pubkey::new_unique());
         index.set_storage_offset(too_big);
     }
+
+    #[test]
+    fn test_data_bucket() {
```
[the remark this answers is not in the document]
Comment (HaoranYi) [reply bucket_map/src/index_entry.rs]: yup.

File: bucket_map/src/index_entry.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (HaoranYi) [reply bucket_map/src/index_entry.rs]: yes.

**Comment by HaoranYi** [comment]
https://github.com/HaoranYi/solana/commit/93687568ca3833418e5b31cbc805b4c060a7aa4d
microbenchmark result:
```
test bench_bank_log2_ceiling_int                  ... bench:         936 ns/iter (+/- 87)
test bench_bank_log2_ceiling_float                ... bench:      11,236 ns/iter (+/- 761)        
```

File: bucket_map/src/index_entry.rs (hunk 1 above)
Comment (brooksprumo) [reply bucket_map/src/index_entry.rs]: A little bit later (after your new assert), there's this:
https://github.com/solana-labs/solana/blob/fe17865bd82fd927280e3ae7c02d5782bdd7ba65/bucket_map/src/bucket.rs#L293-L298

Which implies to me that `num_slots` _can_ be zero (which comes from `data.len()`). I don't remember the context off the top of my head what `num_slots == 0` is used for, but I think the current impl _does_ allow `data.len() == 0`.

Comment (HaoranYi) [reply bucket_map/src/index_entry.rs]: yup. I discovered that too.
Also there is a test that catches this too. 
`bucket_map::tests::bucket_map_test_update_to_0_len`.
I have pushed a commit to return 0 for empty slice, which keeps the behavior the same as before.

File: bucket_map/src/index_entry.rs [hunk 3]
```diff
@@ -153,4 +164,23 @@ mod tests {
         let mut index = IndexEntry::new(Pubkey::new_unique());
         index.set_storage_offset(too_big);
     }
+
+    #[test]
+    fn test_data_bucket_from_num_slots() {
```
Comment (brooksprumo) [inline bucket_map/src/index_entry.rs:169]: Can you update the test to include 0 as well?

**Review by brooksprumo — COMMENTED** [review]
Nice! Can you re-do the benchmark for the new impl?

**Comment by brooksprumo** [comment]
Here's two other impls:
https://godbolt.org/z/xfEsMn11q

File: bucket_map/src/index_entry.rs (hunk 3 above)
Comment (HaoranYi) [reply bucket_map/src/index_entry.rs:169]: done.

**Comment by HaoranYi** [comment]
This is the updated benchmark with the check of empty.
```
test bench_bank_log2_ceiling_int              ... bench:       1,077 ns/iter (+/- 61)
```

**Review by brooksprumo — APPROVED** [review]
lgtm

**Pull request merged**